### ISSUE TITLE
For two different accuracy,which one to choose?

### ISSUE THREAD

**Issue by njoe9** [fchollet/keras#1379]
I'm doing text classification task with keras,but i get two different accuracy,which one should i to choose?
the details as following:
1.the predict method

*Y_result=np.round(np.array(model.predict(X_test, batch_size=batch_size)))
acc = accuracy(Y_test,Y_result)
print('Test accuracy:', acc)*

**the result is about 0.98x.**

2.the evaluate method

*score, acc = model.evaluate(X_test, Y_test, batch_size=batch_size, show_accuracy=True)
print('Test score:', score)
print('Test accuracy:', acc)*

**the result is about 0.77x**

**Now i'm puzzled,which one is the true result?**

3.By the way,another question is word embedding:
*model.add(Embedding(maxfeatures, embedding_size,weights=init_weight))*
For each sentence,i have converted it to many word index,**weights** is word vector matrix.
**My question is how keras to construct word vector for each sentence?
Does keras sum or concatenate all word vectors of a sentence to an output vector?**

Thanks.

**Comment by fchollet** [comment]
Can you post a code snippet to reproduce these results? This will give me a better idea of what is going on.

**Comment by njoe9** [comment]
@fchollet I post the code and output,please look at the following.
1.**CODE**
```
from __future__ import absolute_import
from __future__ import print_function

import pandas as pd 
import numpy as np 

from keras.preprocessing import sequence
from keras.utils.np_utils import accuracy
from keras.optimizers import SGD, RMSprop, Adagrad
from keras.utils import np_utils
from keras.models import Sequential
from keras.layers.core import Dense, Dropout, Activation
from keras.layers.embeddings import Embedding
from keras.layers.recurrent import LSTM, GRU
from keras.models import model_from_json


trasamp=pd.read_excel('train_file.xls',header=None,index=None)
tralen=len(trasamp) 
print("The num of train examples:",tralen)
 
cw = lambda x: list(jieba.cut(x))
trasamp['words'] = trasamp[1].apply(cw)
 
tstsamp =pd.read_excel('test_file.xls',header=None,index=None)
tstlen=len(tstsamp) 
print("The num of test examples:",tstlen)

tstsamp['words'] = tstsamp[1].apply(cw)

df = pd.read_csv('bdbk.txt',sep=' ',header=None) 
df['id']=list(range(0,len(df)))
worddict=df.loc[:,[0,'id']]

wd=df.loc[:,range(1,101)]
init_weight_wv=wd.values

word2idx = dict((c, i) for c, i in zip(worddict[0], worddict['id']))
idx2word = dict((i, c) for c, i in zip(worddict[0], worddict['id']))

char_num = len(init_weight_wv)
idx2word[char_num] = u'UNKNOW-WORDS'
word2idx[u'UNKNOW-WORDS'] = char_num

lastrow=np.random.randn(100,)
init_weight_wv=np.vstack((init_weight_wv,lastrow))
 
def getWordEmbedding(wordlist):
    embeddingresult=[]
    for i in range(0,len(wordlist)):
        predict_word_num = []
        for w in wordlist[i]:
            if w in word2idx:
               predict_word_num.append(word2idx[w])
            else:
               predict_word_num.append(word2idx[u'UNKNOW-WORDS'])
        embeddingresult.append(predict_word_num)
    return embeddingresult

trasamp['embedding'] = getWordEmbedding(trasamp['words']) #Get Train Word Embedding
tstsamp['embedding'] = getWordEmbedding(tstsamp['words']) #Get Test Word Embedding
 
maxlen = 50
 
print("Pad sequences (samples x time)")
trasamp['embedding'] = list(sequence.pad_sequences(trasamp['embedding'], maxlen=maxlen))
tstsamp['embedding'] = list(sequence.pad_sequences(tstsamp['embedding'], maxlen=maxlen))
 

xa = np.array(list(trasamp['embedding'])) 
ya = np.array(list(trasamp[0]))

xt = np.array(list(tstsamp['embedding'])) 
yt = np.array(list(tstsamp[0]))

print('X_train shape:', xa.shape)
print('X_test shape:', xt.shape)

nb_classes=33
print("Convert class vector to binary class matrix (for use with categorical_crossentropy)")
Y_train = np_utils.to_categorical(ya, nb_classes)
Y_test = np_utils.to_categorical(yt, nb_classes)
print('Y_train shape:', Y_train.shape)
print('Y_test shape:', Y_test.shape)

batch_size=8
nb_epoch=10

print('Build model...')
init_weight = [np.array(init_weight_wv)]
maxfeatures = init_weight[0].shape[0] # size of dictionary

model = Sequential()
model.add(Embedding(maxfeatures, 100,weights=init_weight))
model.add(LSTM(256,input_dim=100))
model.add(Dropout(0.5))
model.add(Dense(nb_classes, activation='softmax'))
    
sgd = SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True) # sgd
model.compile(loss='categorical_crossentropy', optimizer=sgd, class_mode="categorical")
 
def trainMode():    
     model.fit(xa, Y_train, batch_size=batch_size,        nb_epoch=nb_epoch,shuffle=True,verbose=1,validation_data=(xt, Y_test),show_accuracy=True)
    
     model.save_weights('lstm_weights_wordemb.h5',overwrite=True) 

def loadMode():
    model.load_weights('lstm_weights_wordemb.h5')
    
def getF1Score(y_t,y_p):
    #[0]:TP,[1]:FP,[2]:FN,[3]:P,[4]:R,[5]:F1
    resultmatrix=np.zeros((33,6))
    
    totalmatrix=np.zeros(6)
    
    for a,b in zip(y_t,y_p):
        for i in range(1,33):
            if a[i]==1:
                if b[i]==1:
                    resultmatrix[i][0]=resultmatrix[i][0]+1#TP+1
                    
                    totalmatrix[0]=totalmatrix[0]+1
                else:
                    resultmatrix[i][2]=resultmatrix[i][2]+1#FN+1
                    
                    totalmatrix[2]=totalmatrix[2]+1
                continue#next
            else:
                if b[i]==1:
                    resultmatrix[i][1]=resultmatrix[i][1]+1#FP+1
                    
                    totalmatrix[1]=totalmatrix[1]+1
                    continue#next

    for i in range(1,33):
       resultmatrix[i][3]= resultmatrix[i][0]/( resultmatrix[i][0]+ resultmatrix[i][1])#p
       resultmatrix[i][4]= resultmatrix[i][0]/( resultmatrix[i][0]+ resultmatrix[i][2])#r
       resultmatrix[i][5]=2* resultmatrix[i][3]* resultmatrix[i][4]/( resultmatrix[i][3]+ resultmatrix[i][4])
   
    totalmatrix[3]=totalmatrix[0]/(totalmatrix[0]+totalmatrix[1])#p
    totalmatrix[4]=totalmatrix[0]/(totalmatrix[0]+totalmatrix[2])#r
    totalmatrix[5]=2*totalmatrix[3]*totalmatrix[4]/(totalmatrix[3]+totalmatrix[4])
    
    return resultmatrix,totalmatrix    

trainMode()
#loadMode()
#print('Y_test:',Y_test)
Y_result=np.round(np.array(model.predict(xt,
                                              batch_size=batch_size)))
#print('Y_result:',Y_result)
acc = accuracy(Y_test,Y_result)
print('The result of using accuracy method:', acc)

score, acc = model.evaluate(xt, Y_test, batch_size=batch_size, show_accuracy=True)
print('Test score:', score)
print('The result of using evaluate method:', acc)


eachf1,totalf1=getF1Score(Y_test,Y_result)
print('eachf1:',eachf1)
print('totalf1:',totalf1)
```

2.**Output**
Train on 47850 samples, validate on 15950 samples
Epoch 1/10
47850/47850 [==============================] - 941s - loss: 1.3976 - acc: 0.6001 - val_loss: 0.9687 - val_acc: 0.7173
Epoch 2/10
47850/47850 [==============================] - 964s - loss: 0.8847 - acc: 0.7432 - val_loss: 0.8371 - val_acc: 0.7584
Epoch 3/10
47850/47850 [==============================] - 970s - loss: 0.7201 - acc: 0.7879 - val_loss: 0.7757 - val_acc: 0.7762
Epoch 4/10
47850/47850 [==============================] - 964s - loss: 0.5999 - acc: 0.8230 - val_loss: 0.7732 - val_acc: 0.7779
Epoch 5/10
47850/47850 [==============================] - 944s - loss: 0.5039 - acc: 0.8492 - val_loss: 0.7872 - val_acc: 0.7841
Epoch 6/10
47850/47850 [==============================] - 974s - loss: 0.4190 - acc: 0.8734 - val_loss: 0.8315 - val_acc: 0.7811
Epoch 7/10
47850/47850 [==============================] - 974s - loss: 0.3454 - acc: 0.8949 - val_loss: 0.8615 - val_acc: 0.7838
Epoch 8/10
47850/47850 [==============================] - 977s - loss: 0.2796 - acc: 0.9140 - val_loss: 0.8914 - val_acc: 0.7861
Epoch 9/10
47850/47850 [==============================] - 942s - loss: 0.2242 - acc: 0.9313 - val_loss: 0.9663 - val_acc: 0.7842
Epoch 10/10
47850/47850 [==============================] - 957s - loss: 0.1808 - acc: 0.9436 - val_loss: 1.0284 - val_acc: 0.7824
**The result of using accuracy method: 0.987534910231**
15950/15950 [==============================] - 60s     
Test score: 1.02839218174
**The result of using evaluate method: 0.782382445141**

totalf1: [  1.22430000e+04   2.85400000e+03   3.70700000e+03   8.10955819e-01
   7.67586207e-01   **7.88675234e-01**]

**Comment by amitmanchanda1995** [comment]
I am having the same issue. Could you explain it.

**Comment by Tauranis** [comment]
Me too.
But the difference on my case is far huge.

On evaluate method: 99.73%
On accuracy method: 3.76%

**Comment by fighting41love** [comment]
same issue

**Comment by srijandas07** [comment]
@Tauranis Did you managed to fix your problem?

**Comment by Tauranis** [comment]
@srijandas07 Yes.

**Comment by srijandas07** [comment]
@Tauranis Could you tell me how did you fix it, and what was the problem?
For me I was using the default accuracy method while evaluating which was giving 90 % accuracy on my test data. But then I noticed that all my labels were predicted wrong. So, then I used all the accuracy methods on keras and concluded that, I was using binary_accuracy by default but I should have used categorical accuracy..

**Comment by a11apurva** [comment]
Facing similar issue. Can someone please let me know how was this fixed??
Thanks..

**Comment by srijandas07** [comment]
Check your loss function!!!
Can you elaborate your problem a bit again!!
